packages/proxy/lib/http/util/service-worker-injector.ts

```diff
@@ -0,0 +1,228 @@
+        }
+
+        // call the original listener
+        const returnValue = listener(event)
+
+        // call the CDP binding to inform the backend whether or not the service worker handled the request
+        sendEvent({ type: 'fetchEvent', payload: { url: event.request.url, isControlled: respondWithCalled } })
+
+        return returnValue
+      }
+    }
+
+    const oldAddEventListener = self.addEventListener
+
+    // Overwrite the addEventListener method so we can determine if the service worker handled the request
+    // https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener
+    self.addEventListener = (type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions) => {
+      if (type === 'fetch' && isValidListener(listener) && !isAborted(options)) {
+        const capture = getCaptureValue(options)
+        const existingListener = capture ? _captureListenersMap.get(listener) : _nonCaptureListenersMap.get(listener)
+
+        // If the listener is already in the map, we don't need to wrap it again
+        if (existingListener) {
+          return oldAddEventListener(type, existingListener, options)
+        }
+
+        let newListener
+
+        // If the listener is a function, we can just wrap it
+        // Otherwise, we need to wrap the listener in a proxy so we can track and wrap the handleEvent function
+        if (typeof listener === 'function') {
+          newListener = __cypressWrapListener(listener)
+        } else {
+          // since the handleEvent function could change, we need to use a proxy to wrap it
+          newListener = new Proxy(listener, {
+            get (target, key) {
+              if (key === 'handleEvent') {
+                const wrappedHandleEvent = _targetToWrappedHandleEventMap.get(target)
+                const origHandleEvent = _targetToOrigHandleEventMap.get(target)
+
+                // If the handleEvent function has not been wrapped yet, or if it has changed, we need to wrap it
+                if ((!wrappedHandleEvent && target.handleEvent) || target.handleEvent !== origHandleEvent) {
+                  _targetToWrappedHandleEventMap.set(target, __cypressWrapListener(target.handleEvent))
+                  _targetToOrigHandleEventMap.set(target, target.handleEvent)
+                }
+
+                return _targetToWrappedHandleEventMap.get(target)
+              }
+
+              return Reflect.get(target, key)
+            },
+          })
+        }
+
+        // get the capture value so we know which map to add the listener to
+        // so we can then remove the listener later if requested
+        getCaptureValue(options) ? _captureListenersMap.set(listener, newListener) : _nonCaptureListenersMap.set(listener, newListener)
+        _listenerCount++
+
+        __cypressServiceWorkerSendHasFetchEventHandlers()
+
+        return oldAddEventListener(type, newListener, options)
```

Comment:
Would it be worth calling this first before our logic? That way if it throws an exception for some reason we won't inadvertently do our logic?

Reply:
good idea [d9a876d](https://github.com/cypress-io/cypress/pull/28900/commits/d9a876ddeb1d2d9da546f506ba9fec3adebacba9)